plugins/channel-safely.cpp

```diff
@@ -0,0 +1,480 @@
+        }
+    }
+    namespace EM = EventManager;
+    if(!has_when<EM::EventHandler>::value){
+        EM::EventHandler tickHandler(onNewDay, dayTicks);
+        EM::registerTick(tickHandler, dayTicks, plugin_self);
+    }
+}
+
+void onStart(color_ostream &out, void* job_ptr){
+    if(enabled && World::isFortressMode()) {
+        auto job = (df::job*) job_ptr;
+        if (is_dig(job) || is_channel(job)) {
+            df::coord local(job->pos);
+            local.x = local.x % 16;
+            local.y = local.y % 16;
+            df::coord above(job->pos);
+            above.z++;
+            df::map_block* block = Maps::getTileBlock(job->pos);
+            //postpone job if above isn't done
+            manager.manage_safety(out, block, local, job->pos, above);
+        }
+    }
+}
+
+void onComplete(color_ostream &out, void* job_ptr) {
+    if (enabled && World::isFortressMode()) {
+        auto job = (df::job*) job_ptr;
+        if (is_channel(job)) {
+            df::coord local(job->pos);
+            local.x = local.x % 16;
+            local.y = local.y % 16;
+            df::coord below(job->pos);
+            below.z--;
+            df::map_block* block = Maps::getTileBlock(below);
+            //activate designation below if group is done now, postpone if not
+            manager.mark_done(job->pos);
+            manager.manage_safety(out, block, local, below, job->pos);
+            manageNeighbours(out, job->pos);
+        }
+    }
+}
+
+void ChannelManager::manage_designations(color_ostream &out, bool full_scan) {
+    if (World::isFortressMode()) {
+        if (full_scan) {
+            out.print("Work safe inspection! Now checking for unsafe work conditions..\n");
+        }
+        static std::once_flag getMapSize;
+        static uint32_t t1, t2, zmax;
+        std::call_once(getMapSize, Maps::getSize, t1, t2, zmax);
+        //debug_out = &out;
+        build_groups(full_scan);
+        debug_out = nullptr;
+        for (auto &group : groups) {
+            for (auto &tile : group) {
+                const df::coord &world_pos = tile.first;
+                df::map_block* block = tile.second;
+                df::coord local(world_pos);
+                local.x = local.x % 16;
+                local.y = local.y % 16;
+                if (world_pos.z < zmax - 1) {
+                    df::coord above(world_pos);
+                    above.z++;
+                    manage_safety(out, block, local, world_pos, above);
+                } else {
+                    block->occupancy[local.x][local.y].bits.dig_marked = false;
+                }
+            }
+        }
+    }
+}
+
+void ChannelManager::manage_safety(color_ostream &out, df::map_block* block, const df::coord &local, const df::coord &tile, const df::coord &tile_above) {
+    df::tile_occupancy &tile_occupancy = block->occupancy[local.x][local.y];
+    for (df::block_square_event* event : block->block_events) {
+        switch (event->getType()) {
```

Comment:
With a C cast, I presume it should still be a "valid" pointer

Reply:
Not sure it would simpler. Wouldn't this be the same as a dynamic cast? it needs to deduce the correct type? Isn't that slow? It's been a while since I decided to avoid cast functions unless built-in to the framework.

Reply:
Using `virtual_cast` helps to futureproof this code against the possibility that Toady changes something with the underlying type hierarchy, or that we find an error in our reverse-engineering. Both of these things happen from time to time.